adds simple logic that keeps authored captions in galleries

draganescu opened the pull request:
## Description
Attempted fix for #8310 

## How has this been tested?
- add a gallery to a post
- edit captions of images
- edit the gallery
- the captions should remain
- delete a caption or more
- edit the gallery and add captions to the files which dont have captions
- the images with empty captions should get the caption from the gallery

## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

draganescu committed:
adds simple logic that keeps authored captions in galleries

draganescu commented:
This is not the best fix yet, but it is a quick fix. @noisysocks you had been bitten by the disappearing captions, do you think this approach could work, and later move to calling the REST media api and update the caption for each attachment based on the GTB gallery 's rich text value?

noisysocks commented on packages/block-library/src/gallery/edit.js:
```diff
@@ -105,11 +105,22 @@ class GalleryEdit extends Component {
+					}
+				}
+				return newImage;
+			} ),
```
We could use [`_.groupBy`](https://devdocs.io/lodash~4/index#groupBy) and [`_.get`](https://devdocs.io/lodash~4/index#get) to make this run a little faster and read more declaratively. 

```js
const { columns, images } = this.props.attributes;

const imagesById = groupBy( images, 'id' );

this.setAttributes( {
	images: newImages.map( ( image ) => ( {
		...pickRelevantMediaFiles( image ),
		caption: get( imageById, [ image.id, 'caption' ], image.caption ),
	} ),
	columns: columns ? Math.min( newImages.length, columns ) : columns,
} );

draganescu replied:
yep, updated :)

noisysocks reviewed (commented):
This makes sense to me. I don't think we need to bother with adding a new REST request. Are there any fields other fields other than `caption` that we should source from the existing selected images?

kjellr commented:
This definitely works better, in that there are no lost captions. That's a massive win on its own.  🙌 

There's still a couple weird experiences that occur under the following conditions: 

1. Insert a gallery, add a caption to (at least) one image. 
2. Hit the "Edit" icon to open the media modal. 
3. Add a new, different caption for that image. Hit "Update Gallery"
4. Notice that the new caption you added is not present, since it's overwritten by the one entered in Gutenberg. 

It'd be great if we could offer logic to see if the caption has been changed since the Media modal was opened, and then update the caption in Gutenberg to match. Not sure if that's possible though. 

And also: 

1. Insert a gallery, include an image with an existing caption. 
2. In Gutenberg, remove the existing caption. 
3. Hit the "Edit" icon to open the media modal. 
4. Do not make changes, but hit "Update Gallery" to return the Gallery block.
5. Notice that the original caption from step 1 is now present again. 

If you modify (not delete) the caption in Gutenberg, the caption will not be overwritten by the one in the media modal. When comparing to that, the behavior in step 5 is inconsistent. I'd expect the editor to know that if I've manually deleted a caption, it's intentionally gone and should not be replaced so easily. 

In general though, I'm super-glad that this means no more lost captions, so it's a definite improvement. 👍

draganescu committed:
adds caption updating if changed in gallery, refactores some code

draganescu commented:
I've updated the code to @noisysocks 's review and also @kjellr I have _added_ the following behaviour:

- if you edit the caption in the gallery modal it will update the one in GTB

Feels a lot better now, let me know if you can take it for another spin :)

draganescu commented:
@gziolo can you please shine some light here as I can't seem to be able to figure out what is wrong with this failing test:

`npm run test-unit -- --testPathPattern test/integration/full-content/full-content.spec.js`

The only thing that causes it to fail appears to be the new item in block.js (attachmentCaption). All the other code passes all tests (I tested by applying each change incrementally).

As soon as I add

```
	"attachmentCaption": {
		"type": "string"
	}
```

to the block.json the test fails. The error is:

```
● full post content fixture › core__gallery

    expect(jest.fn()).not.toHaveErrored(expected)

    Expected mock function not to be called but it was called with:
    ["Unknown source type \"undefined\""]

      32 | 	function assertExpectedCalls() {
      33 | 		if ( spy.assertionsNumber === 0 && spy.mock.calls.length > 0 ) {
    > 34 | 			expect( console ).not[ matcherName ]();
         | 			^
      35 | 		}
      36 | 	}
      37 |

      at Object.assertExpectedCalls (packages/jest-console/src/index.js:34:4)

  ● full post content fixture › core__gallery__columns

    expect(jest.fn()).not.toHaveErrored(expected)

    Expected mock function not to be called but it was called with:
    ["Unknown source type \"undefined\""]

      32 | 	function assertExpectedCalls() {
      33 | 		if ( spy.assertionsNumber === 0 && spy.mock.calls.length > 0 ) {
    > 34 | 			expect( console ).not[ matcherName ]();
         | 			^
      35 | 		}
      36 | 	}
      37 |

      at Object.assertExpectedCalls (packages/jest-console/src/index.js:34:4)
```


If you test be sure to run `npm run test-unit -- --testPathPattern --clearCache test/integration/full-content/full-content.spec.js` to make sure you don't get false positives.

gziolo commented:
> As soon as I add
> 
> ```
> 	"attachmentCaption": {
> 		"type": "string"
> 	}
> ```

Isn't it missing `source` and `attribute` properties? See how the parent `source` with value `query` should be structured: https://github.com/WordPress/gutenberg/blob/master/docs/designers-developers/developers/block-api/block-attributes.md#query

draganescu committed:
refactor code so as to avoig adding data to the DOM and updated the fixtures

draganescu commented:
I have refactored the code and now the failing test works. 

Instead of adding a caption attribute to each image in the gallery I added an attachmentCaptions array attribute of the gallery ( props @talldan ) and  that avoided the need to specify a selector and source.

talldan commented on packages/block-library/src/gallery/block.json:
```diff
@@ -50,6 +50,10 @@
 			"default": "none"
+		},
+		"attachmentCaptions": {
+			"type": "array",
```
I wonder if it'd be better to structure it as an object that maps from the imageId to the attachmentCaption. That way there wouldn't be concerns about keeping the array in order if a new image is inserted into the gallery.

draganescu replied:
it is an array of objects:

```
attachmentCaptions: newImages.map( ( newImage ) => ( {
	id: newImage.id,
	caption: newImage.caption,
} ) ),
```

we already don't care about the image order.

draganescu committed:
updated new deprecated fixture

draganescu committed:
removes manual fixture edits

gziolo commented on packages/block-library/src/gallery/edit.js:
```diff
@@ -125,11 +125,42 @@ class GalleryEdit extends Component {
+				...pickRelevantMediaFiles( newImage ),
+				caption: this.selectCaption( newImage, images, attachmentCaptions ),
+			} ) ),
+			attachmentCaptions: newImages.map( ( newImage ) => ( {
```
We discussed the current solution during the pairing session earlier today. Randomly failing tests on Travis and in local env are a bit puzzling. There is this question whether those captions could be stored in component's state instead?

draganescu replied:
Moved it to state instead of attributes and so the need to edit block.json is gone. Let's see how this tests @gziolo

draganescu committed:
moves attachment captions to state

gziolo commented:
I discovered one more issue when testing it. Steps to reproduce:
- add captions in Media library
- update captions in Gallery
- save and reload
- open Media library for the gallery and click Update Gallery
- captions get replaced even don't modified

![gallery-media](https://user-images.githubusercontent.com/699132/58966195-1fdcc000-87b2-11e9-95b8-2e8baf250452.gif)

draganescu committed:
fixed a problem caused by rebasing

draganescu commented:
@gziolo the bug you found [here](https://github.com/WordPress/gutenberg/pull/15004#issuecomment-499117228) is not really solvable without tying up the gallery component to the media gallery in WP, as for fixing that we need to monitor the state of the attachments in the modal and only update the component if that state is "dirty".

I would push forward with things as they are, what do you think?

gziolo reviewed (approved):
It's better than it used to be. Let's move forward 👍
